File: LC322_CoinChange/dp.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    // Dynamic Programming
    // Time Complexity: O(n)
    // Space Complexity: O(n)
    int coinChange(vector<int>& coins, int amount) {
        if(coins.size() == 0 && amount != 0){
            return -1;
        }
        vector<int> dp(amount+1, amount+1);
        dp[0] = 0;

        for(int i = 1; i < dp.size(); i++){
            for(int coin : coins){
                if(i - coin >= 0){
                    dp[i] = min(dp[i], 1+dp[i-coin]);
                }
            }
        }

        if(dp[amount] == amount+1){
            // There is no suitable combinations
            return -1;
        }
        return dp[amount];
    }
};

File: LC322_CoinChange/dp.py
from typing import List


class Solution:
    # Dynamic Programming
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def coinChange(self, coins: List[int], amount: int) -> int:
        if len(coins) == 0 and amount != 0:
            return -1
        dp = [amount+1]*(amount+1)
        dp[0] = 0

        for i in range(1, len(dp)):
            for coin in coins:
                if i - coin < 0:
                    continue
                else:
                    dp[i] = min(1+dp[i-coin], dp[i])

        return dp[amount] if dp[amount] != amount + 1 else -1
